File: backend/src/graduate_entrance/schemas/vocab.py
from __future__ import annotations

from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

VocabGrade = Literal["forgot", "vague", "mastered"]


class VocabWordRead(BaseModel):
    id: UUID
    word: str
    meaning: str
    phonetic: str
    example_en: str
    example_zh: str
    book_page: int
    ef: float
    interval_days: int
    due_date: date | None
    reps: int


class VocabTodayResponse(BaseModel):
    date: date
    due_words: list[VocabWordRead]
    new_words: list[VocabWordRead]
    due_count: int
    learned_count: int
    total_count: int
    reviewed_today_count: int = Field(
        default=0, description="last_reviewed_on == date 的词数"
    )
    dictation_total_today: int = Field(default=0, description="当日默写总词数")
    dictation_correct_today: int = Field(default=0, description="当日默写写对词数")


class VocabGradeRequest(BaseModel):
    grade: VocabGrade
    as_of: date | None = None


class VocabGradeResult(BaseModel):
    word: VocabWordRead
    grade: VocabGrade
    due_date: date


class VocabDictationResponse(BaseModel):
    date: date
    words: list[VocabWordRead]


class VocabDictationResultRequest(BaseModel):
    correct_word_ids: list[UUID] = Field(default_factory=list)
    wrong_word_ids: list[UUID] = Field(default_factory=list)
    as_of: date | None = None


class VocabDictationResultRead(BaseModel):
    date: date
    total: int
    correct: int


class VocabBulkEnrichStatus(BaseModel):
    running: bool
    processed: int
    failed: int
    remaining: int


class VocabStatsResponse(BaseModel):
    total_count: int
    learned_count: int
    due_count: int
    mastered_count: int = Field(description="reps >= 3 的词数")
    dictation_total_today: int = Field(default=0, description="当日默写总词数")
    dictation_correct_today: int = Field(default=0, description="当日默写写对词数")

File: web/src/services/vocab.ts
import { apiFetch } from './api'

export type VocabGrade = 'forgot' | 'vague' | 'mastered'

export interface VocabWord {
  id: string
  word: string
  meaning: string
  book_page: number
  ef: number
  interval_days: number
  due_date: string | null
  reps: number
}

export interface VocabTodayResponse {
  date: string
  due_words: VocabWord[]
  new_words: VocabWord[]
  due_count: number
  learned_count: number
  total_count: number
}

export interface VocabGradeResult {
  word: VocabWord
  grade: VocabGrade
  due_date: string
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await apiFetch(`/api${path}`, {
    ...init,
    headers: {
      'Content-Type': 'application/json',
      ...init?.headers,
    },
  })
  if (!response.ok) {
    throw new Error(`Vocab API 请求失败：${response.status}`)
  }
  return (await response.json()) as T
}

export function fetchVocabToday(newLimit = 50): Promise<VocabTodayResponse> {
  return request<VocabTodayResponse>(`/vocab/today?new_limit=${newLimit}`)
}

export function gradeVocabWord(wordId: string, grade: VocabGrade): Promise<VocabGradeResult> {
  return request<VocabGradeResult>(`/vocab/${wordId}/grade`, {
    method: 'POST',
    body: JSON.stringify({ grade }),
  })
}
